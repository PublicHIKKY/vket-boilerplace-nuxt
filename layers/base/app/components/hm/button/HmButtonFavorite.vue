<template>
  <div
    class="hm-button-favorite"
    :class="{ '-disabled': disabled }"
  >
    <div
      class="button"
      @click="onClick"
    >
      <IconFavorite
        class="favorite-icon"
        :class="{ '-active': value, '-disabled': disabled }"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import IconFavorite from '#base/app/assets/icons/icon-heart.svg?component'

const props = withDefaults(
  defineProps<{
    value: boolean
    disabled?: boolean
  }>(),
  {
    disabled: false,
  },
)
const emit = defineEmits<{
  (event: 'input', value: boolean): void
}>()

const onClick = () => {
  if (props.disabled) return
  emit('input', !props.value)
}
</script>

<style lang="scss" scoped>
@use '#base/app/assets/styles/variables' as v;

.hm-button-favorite {
  > .button {
    cursor: pointer;
    height: 24px;
    width: 24px;
  }

  &.-disabled {
    > .button {
      cursor: not-allowed;
    }
  }

  .button:deep(.body) {
    fill: none;
  }

  .button:deep(.border) {
    fill: #f5f5f5;
  }
}

.favorite-icon:not(.-disabled) {
  filter: drop-shadow(0 0 3px rgba(#fff, 0.75));

  .body {
    fill: #fff;
    transition: fill 0.1s ease-in;
  }

  &.-active,
  &:hover {
    .border,
    .body {
      fill: #f2509c;
    }
  }
}

.favorite-icon.-disabled {
  .body {
    fill: v.$gray-3;
    transition: fill 0.1s ease-in;
  }
}
</style>
