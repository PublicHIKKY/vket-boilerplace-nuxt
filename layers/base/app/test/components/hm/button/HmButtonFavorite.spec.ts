import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import HmButtonFavorite from '#base/app/components/hm/button/HmButtonFavorite.vue'

test('ref component', () => {
  expect(HmButtonFavorite).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HmButtonFavorite, { props: { value: true } })
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

describe('props', () => {
  describe(':value', () => {
    test('is active', () => {
      const wrapper = mount(HmButtonFavorite, { props: { value: true } })
      expect(wrapper.get('.favorite-icon').attributes('class')).toBe(
        'favorite-icon -active',
      )
    })

    test('is not active', () => {
      const wrapper = mount(HmButtonFavorite, { props: { value: false } })
      expect(wrapper.get('.favorite-icon').attributes('class')).toBe(
        'favorite-icon',
      )
    })
  })

  describe(':disabled', () => {
    test('default is false', () => {
      const wrapper = mount(HmButtonFavorite, { props: { value: true } })
      expect(wrapper.get('.hm-button-favorite').attributes('class')).toBe(
        'hm-button-favorite',
      )
      expect(wrapper.get('.favorite-icon').attributes('class')).toBe(
        'favorite-icon -active',
      )
    })

    test('is disabled', () => {
      const wrapper = mount(HmButtonFavorite, {
        props: { value: true, disabled: true },
      })
      expect(wrapper.get('.hm-button-favorite').attributes('class')).toBe(
        'hm-button-favorite -disabled',
      )
      expect(wrapper.get('.favorite-icon').attributes('class')).toBe(
        'favorite-icon -active -disabled',
      )
    })
  })
})

describe('emits', () => {
  test('click emits return false when value is true', async () => {
    const wrapper = mount(HmButtonFavorite, { props: { value: true } })
    await wrapper.get('.button').trigger('click')
    expect(wrapper.emitted('input')?.[0]).toEqual([false])
  })

  test('click emits return true when value is false', async () => {
    const wrapper = mount(HmButtonFavorite, { props: { value: false } })
    await wrapper.get('.button').trigger('click')
    expect(wrapper.emitted('input')?.[0]).toEqual([true])
  })

  test('click emits no event when disabled', async () => {
    const wrapper = mount(HmButtonFavorite, {
      props: { value: true, disabled: true },
    })
    await wrapper.get('.button').trigger('click')
    expect(wrapper.emitted('input')).toBeFalsy()
  })
})
