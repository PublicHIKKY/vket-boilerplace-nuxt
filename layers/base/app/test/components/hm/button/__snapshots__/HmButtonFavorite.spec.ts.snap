// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `
"<div data-v-f3cd9c7d="" class="hm-button-favorite">
  <div data-v-f3cd9c7d="" class="button"><svg data-v-f3cd9c7d="" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 31.02 27.03" class="favorite-icon -active">
      <defs>
        <style>
          .body {
            fill: none;
          }

          .border {
            fill: #757575;
          }
        </style>
      </defs>
      <g class="body">
        <path d="M28.63,2.41c-3.19-3.18-8.35-3.18-11.54,0l-1.57,1.57-1.57-1.57c-3.17-3.2-8.33-3.22-11.53-.05-3.2,3.17-3.22,8.33-.05,11.53,.01,.01,.03,.03,.04,.04l1.57,1.57,11.54,11.52,11.54-11.52,1.57-1.57c3.18-3.17,3.19-8.33,.01-11.51,0,0,0,0-.01-.01Z"></path>
      </g>
      <g class="border">
        <path d="M22.86,2.03c-1.65,0-3.19,.64-4.36,1.8l-1.57,1.57c-.78,.78-2.05,.78-2.83,0l-1.57-1.57c-1.16-1.16-2.71-1.8-4.36-1.8s-3.19,.64-4.36,1.8c-1.16,1.16-1.8,2.7-1.8,4.35s.64,3.18,1.8,4.35l11.7,11.68,11.7-11.68c1.16-1.16,1.8-2.7,1.8-4.35s-.64-3.18-1.8-4.35c-1.16-1.16-2.71-1.8-4.36-1.8m0-2c2.09,0,4.18,.8,5.77,2.39h0c3.19,3.18,3.19,8.34,0,11.52l-13.11,13.09L2.41,13.94C-.78,10.75-.78,5.6,2.41,2.41,5.59-.77,10.76-.77,13.94,2.41l1.57,1.57,1.57-1.57C18.68,.82,20.77,.03,22.86,.03Z"></path>
      </g>
    </svg></div>
</div>"
`;
